src/query/functions/aggregation/count_values.go

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregation
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+	"github.com/m3db/m3/src/query/ts"
+)
+
+const (
+	// CountValuesType counts the number of non nan elements with the same value
+	CountValuesType = "count_values"
+)
+
+// NewCountValuesOp creates a new count values operation
+func NewCountValuesOp(
+	opType string,
+	params NodeParams,
+) (parser.Params, error) {
+	if opType != CountValuesType {
+		return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	return newCountValuesOp(params, opType), nil
+}
+
+// countValuesOp stores required properties for count values ops
+type countValuesOp struct {
+	params NodeParams
+	opType string
+}
+
+// OpType for the operator
+func (o countValuesOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o countValuesOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o countValuesOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &countValuesNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func newCountValuesOp(params NodeParams, opType string) countValuesOp {
+	return countValuesOp{
+		params: params,
+		opType: opType,
+	}
+}
+
+type countValuesNode struct {
+	op         countValuesOp
+	controller *transform.Controller
+}
+
+// bucketColumn represents a column of times a particular value in a series has
+// been seen. This may expand as more unique values are seen
+type bucketColumn []int
```

Comment:
consider using floats ? int would also be 64 bits. You can also avoid the weird -1 -> NaN conversion

Reply:
https://github.com/m3db/m3/pull/874#discussion_r214414809 :|